### PR TITLE
[FEATURE REQUEST] Add status message when (un)setting av. offline from preview

### PR DESCRIPTION
**Pull request by joragua**
## Related Issues
App: https://github.com/owncloud/android/issues/4382

- [x] Add changelog files for the fixed issues in folder changelog/unreleased. More info [here](https://github.com/owncloud/android/tree/master/changelog#create-changelog-items)
- [x] Add feature to Release Notes in `ReleaseNotesViewModel.kt` creating a new `ReleaseNote()` with String resources (if required)

_____

## QA

Checks done: https://github.com/owncloud/android/pull/4482#issuecomment-2401663910

Reports:

- [X] (1) Snackbar not correctly displayed in image preview https://github.com/owncloud/android/pull/4482#issuecomment-2401746868

### CODE REVIEW

File: owncloudApp/src/main/java/com/owncloud/android/presentation/previews/PreviewTextViewModel.kt [hunk 1]
```diff
@@ -38,11 +40,17 @@ class PreviewTextViewModel(
     private val filterFileMenuOptionsUseCase: FilterFileMenuOptionsUseCase,
     private val contextProvider: ContextProvider,
     private val coroutinesDispatcherProvider: CoroutinesDispatcherProvider,
+    getFileByIdAsStreamUseCase: GetFileByIdAsStreamUseCase,
```
Comment (JuancaG05) [inline owncloudApp/src/main/java/com/owncloud/android/presentation/previews/PreviewTextViewModel.kt:43]: We usually group all the use cases together in the parameters, and then the left ones 👍 (applicable to all ViewModels in which you introduced new parameters)

File: owncloudApp/src/main/java/com/owncloud/android/presentation/previews/PreviewTextViewModel.kt [hunk 2]
```diff
@@ -38,11 +40,17 @@ class PreviewTextViewModel(
...
+    ocFile: OCFile
```
Comment (JuancaG05) [inline owncloudApp/src/main/java/com/owncloud/android/presentation/previews/PreviewTextViewModel.kt:44]: Let's use trailing commas to keep versioning cleaner. More info: https://kotlinlang.org/docs/coding-conventions.html#trailing-commas

File: owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewImageViewModel.kt [hunk 3]
```diff
@@ -45,15 +47,22 @@ class PreviewImageViewModel(
     private val getLiveDataForFinishedDownloadsFromAccountUseCase: GetLiveDataForFinishedDownloadsFromAccountUseCase,
     private val filterFileMenuOptionsUseCase: FilterFileMenuOptionsUseCase,
     private val contextProvider: ContextProvider,
-    private val coroutinesDispatcherProvider: CoroutinesDispatcherProvider
+    private val coroutinesDispatcherProvider: CoroutinesDispatcherProvider,
+    getFileByIdAsStreamUseCase: GetFileByIdAsStreamUseCase,
+    ocFile: OCFile,
 ) : ViewModel() {
 
+
```
Comment (JuancaG05) [inline owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewImageViewModel.kt:55]: Why this new line? 🤠

File: owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewAudioFragment.kt [hunk 4]
```diff
@@ -285,11 +300,13 @@ class PreviewAudioFragment : FileFragment() {
 
             R.id.action_set_available_offline -> {
                 fileOperationsViewModel.performOperation(FileOperation.SetFilesAsAvailableOffline(listOf(file)))
+                Snackbar.make(requireView(), R.string.sync_file_nothing_to_do_msg, Snackbar.LENGTH_LONG).show()
```
Comment (JuancaG05) [inline owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewAudioFragment.kt:303]: Uhm... not sure if showing the message `File contents already synchronized` when pressing on both `Set/Unset as available offline` options is suitable. This deserves another iteration, maybe with more specific messages. Same in the other preview classes

File: owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewAudioFragment.kt [hunk 5]
```diff
@@ -115,6 +120,16 @@ class PreviewAudioFragment : FileFragment() {
     override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
         super.onViewCreated(view, savedInstanceState)
 
+        collectLatestLifecycleFlow(previewAudioViewModel.getCurrentFile()) { currentFile ->
+            if (currentFile != null) {
+                file = currentFile
+                requireActivity().invalidateOptionsMenu()
```
Comment (JuancaG05) [inline owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewAudioFragment.kt:126]: Just to confirm: by doing this here, we don't need to change current references to `file` in the `onPrepareOptionsMenu` and `onOptionsItemSelected` methods, right?

File: owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewImageFragment.kt [hunk 6]
```diff
@@ -332,6 +347,14 @@ class PreviewImageFragment : FileFragment() {
         return if (isSVGFile(file)) Color.WHITE else Color.BLACK
     }
 
+    private fun showSnackbarMessage() {
+        val snackbar = Snackbar.make(requireView(), R.string.sync_file_nothing_to_do_msg, Snackbar.LENGTH_LONG)
+        val params = snackbar.view.layoutParams as ViewGroup.MarginLayoutParams
+        params.setMargins(params.leftMargin, params.topMargin, params.rightMargin, 100)
```
Comment (JuancaG05) [inline owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewImageFragment.kt:353]: What happens if we don't set these margins for the Snackbar? Is it special about the image preview?

File: owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewVideoActivity.kt [hunk 7]
```diff
@@ -80,8 +83,10 @@ import kotlinx.coroutines.flow.Flow
 import kotlinx.coroutines.flow.collectLatest
 import kotlinx.coroutines.launch
 import org.koin.androidx.viewmodel.ext.android.viewModel
+import org.koin.core.parameter.parametersOf
 import timber.log.Timber
 
+@RequiresApi(Build.VERSION_CODES.TIRAMISU)
```
Comment (JuancaG05) [inline owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewVideoActivity.kt:89]: Hmmm I don't like this, this way you're restricting this class to SDK 33+ (Android 13+) devices. Why is that needed?

File: owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewVideoActivity.kt [hunk 8]
```diff
@@ -92,7 +97,7 @@ class PreviewVideoActivity : FileActivity(), Player.Listener, OnPrepareVideoPlay
     private var playWhenReady = true
     private var playbackPosition: Long = 0
     private var windowInsetsController: WindowInsetsControllerCompat? = null
-    private val previewVideoViewModel: PreviewVideoViewModel by viewModel()
+    private val previewVideoViewModel: PreviewVideoViewModel by viewModel() { parametersOf(intent.getParcelableExtra(EXTRA_FILE, OCFile::class.java)) }
```
Comment (JuancaG05) [inline owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewVideoActivity.kt:100]: Can't you extract it from `arguments` in this case? Is it necessary to explicitly indicate the class type as second parameter?

File: owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewVideoActivity.kt [hunk 9]
```diff
@@ -318,6 +333,8 @@ class PreviewVideoActivity : FileActivity(), Player.Listener, OnPrepareVideoPlay
     }
 
     override fun onPrepareMenu(menu: Menu) {
+
+
```
Comment (JuancaG05) [inline owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewVideoActivity.kt:337]: Again these new blank lines why? 😁

**Review by JuancaG05 — CHANGES_REQUESTED** [review]
Good job @joragua! Some comments and questions here 👍

File: owncloudApp/src/main/java/com/owncloud/android/presentation/previews/PreviewTextViewModel.kt (hunk 1 above)
Comment (joragua) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/previews/PreviewTextViewModel.kt:43]: Changed 👍

File: owncloudApp/src/main/java/com/owncloud/android/presentation/previews/PreviewTextViewModel.kt (hunk 2 above)
Comment (joragua) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/previews/PreviewTextViewModel.kt:44]: Done!

File: owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewImageViewModel.kt (hunk 3 above)
Comment (joragua) [reply owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewImageViewModel.kt:55]: Removed ✅

File: owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewVideoActivity.kt (hunk 9 above)
Comment (joragua) [reply owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewVideoActivity.kt:337]: Removed too. I don't know why that lines are there. 😅

File: owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewAudioFragment.kt (hunk 4 above)
Comment (joragua) [reply owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewAudioFragment.kt:303]: I've used that string because it is used in file details screen. It can be used `The file was set as (un)available offline correctly` string instead. Also, it can be changed in `FileDetailsFragment` because it uses the same string.

File: owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewAudioFragment.kt (hunk 5 above)
Comment (joragua) [reply owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewAudioFragment.kt:126]: Yes, that's right. Te value of `currentFile` is in the variable called `file`, so we don't need to change te reference there. This is the same with all previews. 😃

File: owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewVideoActivity.kt (hunk 7 above)
Comment (joragua) [reply owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewVideoActivity.kt:89]: This is necessary if `getParcelableExtra` method has a second parameter, but I've changed it and that annotation has been removed. ✅

File: owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewVideoActivity.kt (hunk 8 above)
Comment (joragua) [reply owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewVideoActivity.kt:100]: Changed. It's no necessary the second parameter.

File: owncloudApp/src/main/java/com/owncloud/android/presentation/previews/PreviewAudioViewModel.kt [hunk 10]
```diff
@@ -37,11 +39,17 @@ class PreviewAudioViewModel(
     private val filterFileMenuOptionsUseCase: FilterFileMenuOptionsUseCase,
     private val contextProvider: ContextProvider,
     private val coroutinesDispatcherProvider: CoroutinesDispatcherProvider,
+    getFileByIdAsStreamUseCase: GetFileByIdAsStreamUseCase,
```
Comment (JuancaG05) [inline owncloudApp/src/main/java/com/owncloud/android/presentation/previews/PreviewAudioViewModel.kt:42]: You didn't group use cases in parameters here

File: owncloudApp/src/main/java/com/owncloud/android/presentation/previews/PreviewAudioViewModel.kt [hunk 11]
```diff
@@ -37,11 +39,17 @@ class PreviewAudioViewModel(
...
+    ocFile: OCFile
```
Comment (JuancaG05) [inline owncloudApp/src/main/java/com/owncloud/android/presentation/previews/PreviewAudioViewModel.kt:43]: You didn't add trailing comma here

File: owncloudApp/src/main/java/com/owncloud/android/presentation/previews/PreviewVideoViewModel.kt [hunk 12]
```diff
@@ -37,11 +39,17 @@ class PreviewVideoViewModel(
     private val filterFileMenuOptionsUseCase: FilterFileMenuOptionsUseCase,
     private val contextProvider: ContextProvider,
     private val coroutinesDispatcherProvider: CoroutinesDispatcherProvider,
+    getFileByIdAsStreamUseCase: GetFileByIdAsStreamUseCase,
+    ocFile: OCFile
```
Comment (JuancaG05) [inline owncloudApp/src/main/java/com/owncloud/android/presentation/previews/PreviewVideoViewModel.kt:43]: Same as in `PreviewAudioViewModel`

File: owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewImageViewModel.kt [hunk 13]
```diff
@@ -45,7 +47,9 @@ class PreviewImageViewModel(
     private val getLiveDataForFinishedDownloadsFromAccountUseCase: GetLiveDataForFinishedDownloadsFromAccountUseCase,
     private val filterFileMenuOptionsUseCase: FilterFileMenuOptionsUseCase,
     private val contextProvider: ContextProvider,
-    private val coroutinesDispatcherProvider: CoroutinesDispatcherProvider
+    private val coroutinesDispatcherProvider: CoroutinesDispatcherProvider,
+    getFileByIdAsStreamUseCase: GetFileByIdAsStreamUseCase,
```
Comment (JuancaG05) [inline owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewImageViewModel.kt:51]: And same here as well

File: owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewAudioFragment.kt [hunk 14]
```diff
@@ -342,6 +359,15 @@ class PreviewAudioFragment : FileFragment() {
...
 
+    private fun showSnackbarMessage(){
+        val snackbarString = if (file.isAvailableOffline){
+            R.string.confirmation_unset_available_offline
+        } else {
+            R.string.confirmation_set_available_offline
+        }
+        Snackbar.make(requireView(), snackbarString, Snackbar.LENGTH_LONG).show()
+    }
+
```
Comment (JuancaG05) [inline owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewAudioFragment.kt:370]: I don't know if it is worth it to separate this in another method when you can do the same in-place with just 1 line and you only call it twice, one per each `if` branch. Moreover, the condition would be the opposite I think: if `file` is available offline is because you just set it (so, you just pressed the `Set as available offline` option and thus the message in the Snackbar should say that, not the opposite)

File: owncloudApp/src/main/res/values/strings.xml [hunk 15]
```diff
@@ -369,6 +369,8 @@
 
     <string name="set_available_offline">Set as available offline</string>
     <string name="unset_available_offline">Unset as available offline</string>
+    <string name="confirmation_set_available_offline">The file was set as available offline correctly</string>
+    <string name="confirmation_unset_available_offline">The file was set as unavailable offline correctly</string>
```
Comment (JuancaG05) [inline owncloudApp/src/main/res/values/strings.xml:373]: The name of the state is `available offline`, so let's not say that a file is set as `unavailable offline`, better say that a file was unset as `available offline`, just as the button in the menu indicates

File: owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewAudioFragment.kt [hunk 16]
```diff
@@ -342,6 +359,15 @@ class PreviewAudioFragment : FileFragment() {
         }
     }
 
+    private fun showSnackbarMessage(){
+        val snackbarString = if (file.isAvailableOffline){
+            R.string.confirmation_unset_available_offline
+        } else {
+            R.string.confirmation_set_available_offline
+        }
```
Comment (JuancaG05) [inline owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewAudioFragment.kt:367]: Use these 2 messages in `FileDetailsFragment` as well

File: owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewVideoActivity.kt [hunk 17]
```diff
@@ -92,7 +96,7 @@ class PreviewVideoActivity : FileActivity(), Player.Listener, OnPrepareVideoPlay
     private var playWhenReady = true
     private var playbackPosition: Long = 0
     private var windowInsetsController: WindowInsetsControllerCompat? = null
-    private val previewVideoViewModel: PreviewVideoViewModel by viewModel()
+    private val previewVideoViewModel: PreviewVideoViewModel by viewModel { parametersOf(intent.getParcelableExtra(EXTRA_FILE)) }
```
Comment (JuancaG05) [inline owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewVideoActivity.kt:99]: In any case, can't you extract the file from `arguments` in this case? Does it need to be from `intent`?

File: owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewVideoActivity.kt [hunk 18]
```diff
@@ -318,6 +332,8 @@ class PreviewVideoActivity : FileActivity(), Player.Listener, OnPrepareVideoPlay
     }
 
     override fun onPrepareMenu(menu: Menu) {
+
+
```
Comment (JuancaG05) [inline owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewVideoActivity.kt:336]: You didn't remove these lines 😞

**Review by JuancaG05 — CHANGES_REQUESTED** [review]
Some more comments and changes here @joragua 🧑🏻‍💻

**Comment by JuancaG05** [comment]
There are several KtLint reports as well in the BitRise workflow, take a look at them and fix them. You can rebase this branch against master as well @joragua 👍

File: owncloudApp/src/main/java/com/owncloud/android/presentation/previews/PreviewAudioViewModel.kt (hunk 10 above)
Comment (joragua) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/previews/PreviewAudioViewModel.kt:42]: Done 👍

File: owncloudApp/src/main/java/com/owncloud/android/presentation/previews/PreviewAudioViewModel.kt (hunk 11 above)
Comment (joragua) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/previews/PreviewAudioViewModel.kt:43]: Added!

File: owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewVideoActivity.kt (hunk 17 above)
Comment (joragua) [reply owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewVideoActivity.kt:99]: It's not possible because it's an `Activity` and `requireArguments()` only exists for `Fragments`

File: owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewImageViewModel.kt (hunk 13 above)
Comment (joragua) [reply owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewImageViewModel.kt:51]: Changed too ✅

File: owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewAudioFragment.kt (hunk 14 above)
Comment (joragua) [reply owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewAudioFragment.kt:370]: I've removed the method and now the `Snackbar` string depends on the condition.

File: owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewVideoActivity.kt (hunk 18 above)
Comment (joragua) [reply owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewVideoActivity.kt:336]: Removed!

File: owncloudApp/src/main/res/values/strings.xml (hunk 15 above)
Comment (joragua) [reply owncloudApp/src/main/res/values/strings.xml:373]: I've changed the string (`The file was unset as available offline correctly`)

File: owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewAudioFragment.kt (hunk 16 above)
Comment (joragua) [reply owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewAudioFragment.kt:367]: I've checked it and the logic in `FileDetailsFragment` is different, so it's not applicable in this case.

File: owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewAudioFragment.kt [hunk 19]
```diff
@@ -285,11 +301,13 @@ class PreviewAudioFragment : FileFragment() {
 
             R.id.action_set_available_offline -> {
                 fileOperationsViewModel.performOperation(FileOperation.SetFilesAsAvailableOffline(listOf(file)))
+                    Snackbar.make(requireView(), R.string.confirmation_set_available_offline, Snackbar.LENGTH_LONG).show()
```
Comment (JuancaG05) [inline owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewAudioFragment.kt:304]: This line has an additional indentation. KtLint won't like that 😅

**Review by JuancaG05 — CHANGES_REQUESTED** [review]
One more tiny change and we're done @joragua!

File: owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewAudioFragment.kt (hunk 19 above)
Comment (joragua) [reply owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewAudioFragment.kt:304]: Removed 🚀

**Review by JuancaG05 — APPROVED** [review]
LGTM now! Good job @joragua 🏅

**Comment by jesmrec** [comment]
Checks done:

1. Set as av. offline

- [ ] Image preview
- [ ] Text preview
- [ ] Audio preview
- [ ] Video preview

2. Unset as av. offline 

- [ ] Image preview
- [ ] Text preview
- [ ] Audio preview
- [ ] Video preview

**Comment by jesmrec** [comment]
### (1)

just a tiny detail i noticed: in the image previews, the snackbar is smaller (in height) than in other previews. I guess that is something related with system, but it took my attention. I tested with two different devices. Check the difference with the following short videos:


|txt preview|image preview|
|:-:|:-:|
|https://github.com/user-attachments/assets/f25a79da-783d-4ad1-9fbb-d580cd838934|https://github.com/user-attachments/assets/9407ce38-ce91-49fc-bb43-1e82cb993bdf|

i suggest to download both of them and play together to compare.

Pixel 2, Android 11
Xiaomi Redmi 13, Android 14
`cdbc8705`

**Comment by jesmrec** [comment]
Related with (1), by setting as av. offline in landscape orientation (image preview):


https://github.com/user-attachments/assets/a05366f9-c359-4c9b-a551-21dc8728fbc9

- The snackbar does not rise from the bottom (maybe because the margin)
- The android bar, placed to the left, overlaps part of the message

could this be part of the same issue?

**Review by JuancaG05 — APPROVED** [review]
LGTM 👍

**Comment by jesmrec** [comment]
(1) fixed

**Comment by jesmrec** [comment]
Approved on my side. 

There are some conflicts in the branch because the changelog items, could you take a look before merging @joragua ?
